Persistence strategy

rafamanzo opened the pull request:
This does more than just adding the Persistence Strategy!

It introduces the interfaces for Parsers and Collector.

(rebased version of https://github.com/mezuro/kolekti/pull/2)

danielkza committed:
Correct requires and add missing module decl. in Entities

It's probably not a good idea to require kalibro_client on the root of
the module for the submodule to use: it's hard to find out where it's
coming from, and it makes the order of module inclusion significant,
which can cause confusion later.

Signed off by: Heitor Reis <marcheing@gmail.com>

danielkza committed:
Import Metric factories from kalibro-processor

Signed off by: Heitor Reis <marcheing@gmail.com>

danielkza committed:
Import Collector from KalibroProcessor and refactor Runner

Runner now just takes the desired metrics and the code path and chooses
the appropriate collectors to call.

Collector methods are now instance methods, which makes it more generic
for clients to decide how their implementations are done. It makes
Runner::BaseMetricRunner unnecessary, and most of the code in
Runner::Main was moved to runner.

Signed off by: Heitor Reis <marcheing@gmail.com>

danielkza committed:
Allow access to currently registered collectors and unregistration

Signed off by: Heitor Reis <marcheing@gmail.com>

danielkza committed:
Require Collector in the main module

Signed off by: Heitor Reis <marcheing@gmail.com>

marcheing committed:
Add persistence strategy for metric results

Signed off by: Daniel Miranda <danielkza2@gmail.com>

danielkza committed:
Require the errors module in the main module

danielkza committed:
Update persistence strategy API

Change the order of some parameters and implement a method for creating
multiple related hotspot results.

danielkza committed:
Add a secondary metric configuration factory

Signed off by: Heitor Reis <marcheing@gmail.com>

danielkza committed:
Fix collector tests

danielkza committed:
Fix runner tests

Signed off by: Heitor Reis <marcheing@gmail.com>

danielkza committed:
Add available_collectors method to Kolekti

Signed off by: Heitor Reis <marcheing@gmail.com>

danielkza committed:
Fix missing module declaratio in errors.rb and create CollectorError

Signed off by: Heitor Reis <marcheing@gmail.com>

diegoamc committed:
Implemented default_value_from method for collectors

Includes a default implementation for Analizo

Signed off by: Daniel Miranda <danielkza2@gmail.com>

danielkza committed:
Implemented default_metric_value in Kolekti package

It will look up the collector for a metric configuration, then grab the
default value from it.

Signed off by: Rafael Reggiani Manzo <rr.manzo@gmail.com>

marcheing committed:
Add factory for persistence strategy

Move the TestPersistenceStrategy out of the class.
Also, not accepting a Hash as a parameter on Runner#initialize for
wanted_metric_configurations.

Signed-off-by: Diego Araújo <diegoamc90@gmail.com>

diegoamc committed:
Include a native metric with SOFTWARE scope on tests

Signed-off-by: Heitor Reis <marcheing@gmail.com>

diegoamc commented on lib/kolekti/collector.rb:
```diff
@@ -0,0 +1,44 @@
+
+    def parse_supported_metrics(metrics_path, metric_collector_name, languages)
+      supported_metrics = {}
+      YAML.load_file(metrics_path)[:metrics].each do | key, value |
```
For better readability, would it not be nice to rename `key` to `code` and `value` to `metric`?

rafamanzo committed:
Rename parse_supported_metrics iterator vars

`key` and `value` are not meaningful names, so we replaced them actual
ones.

rafamanzo committed:
Add unit test coverage for hotspot metric parsing

It was missing this coverage which is necessary and thus the related
factories.

rafamanzo commented:
Just updated the PR with commits fixing the commented and a extra commit regarding unit test coverage.
